docs/create-packages/native-files-in-net-packages.md

```diff
@@ -0,0 +1,88 @@
+.NET Framework projects using packages with `pacakges.config`, rather than `PackageReference`, have no support for the `ref/` and `runtimes/` features described below.
+
+While NuGet has a [`contentFiles/` feature](../reference/msbuild-targets#including-content-in-a-package), it is not recommended to use this, as it doesn't support per-RID selection, and therefore doesn't participate in the .NET runtime's probing paths features.
+
+## Background
+
+First we'll consider a project referencing packages.
+When the app is published for a specific [Runtime Identifier (RID)](/dotnet/core/rid-catalog), then NuGet package assets for that RID will be copied into the output directory.
+When the app is published without a specific RID, all RID specific content will be published in subdirectories, and a `{project}.deps.json` file is created that the .NET runtime loads to determine the .
+
+## Understanding NuGet package asset selection
+
+There are two [types of assets](../consume-packages/Package-References-in-Project-Files.md#controlling-dependency-assets) in a NuGet file that are most relevant to this scenario, compile and runtime assets.
+
+### Compile assets
+
+Compile assets are passed to the compiler, and therefore must only contain .NET assemblies.
+NuGet will select compile assets from `ref/{tfm}/`, and if that directory does not exist, it will fall back to `lib/{tfm}`, where `{tfm}` is the canonical target framework of the project referencing the package.
+
+The .NET compiler will generate warnings when an assembly targets a different architecture than what the current compilation is targeting, so your package consumers will have the best experience if your `ref/` or `lib/` assemblies target `AnyCPU`.
+
+If your assembly contains a lot of managed code, not just P/Invoke method definitions, you can save space by making the `ref/` assembly a [reference assembly](/dotnet/standard/assembly/reference-assemblies#generating-reference-assemblies), but be sure to include it under `ref/` in your package, not `lib/`.
+
+Therefore, there is no capability to provide different compile-time assemblies for different RIDs, you will need to provide the same API surface for all RIDs, and do runtime checks (for example, throw )
+
+### Runtime assets
+
+Runtime assets are copied to the output directory on build and publish, and the .NET SDK generates the `deps.json` with all the RID-specific asset information.
+NuGet will select runtime assets from `runtimes/{rid}/native/` and `runtimes/{rid}/lib/{tfm}/` (see below to understand `sub-path`), and if that doesn't exist, then it will fall back to `lib/{tfm}/`.
+
+When there are both `runtimes/{rid}/native/` and `runtimes/{rid}/lib/{tfm}/` files, NuGet will always select the `runtimes/{rid}/native/` assets, and then the assets from only one of the `runtimes/{rid}/lib/{tfm}/` directories, just like it does for compile assets in `ref/{tfms}/` or `lib/{tfms}/`.
```

Comment:
This should talk about native assets separate from runtime assets - perhaps with better terms.

It might be good to spell out all the different asset types and how NuGet selects them.  These are the asset types exposed: https://github.com/NuGet/NuGet.Client/blob/73da89e27dcb39933133aa056ade47c8678e84a1/src/NuGet.Core/NuGet.LibraryModel/LibraryIncludeFlag.cs#L9

Compile: managed assembly, passed to the compiler -- best ref/tfm (if exists), then best lib/tfm
Runtime: managed assembly, copied at runtime (listed in deps for host to load) -- best runtime/rid/lib/tfm, giving precedence first to RID, then tfm, then best lib/tfm
Native: native library, copied at runtime (listed in deps for host to load) -- best runtime/rid/nativeAssets/tfm, giving precedence first to RID, then tfm, then best runtime/rid/native.
etc..

In all cases nuget selects the best folder, then exposes all the items from that folder (based on a naming convention).

I believe all of this logic is encoded here: https://github.com/NuGet/NuGet.Client/blob/73da89e27dcb39933133aa056ade47c8678e84a1/src/NuGet.Core/NuGet.Packaging/ContentModel/ManagedCodeConventions.cs#L411

Precedence for TFMs can be explored using https://nugettools.azurewebsites.net/6.4.0, precedence for RIDs can be explored using https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.NETCore.Platforms/src/runtime.compatibility.json.

Reply:
Thanks @ericstj, this really helped. I've pushed an update which I hope improves correctness of the docs, and also makes it easier to understand. I actually had no idea that runtime and native were different asset types, despite having debugged `ManagedCodeConventions` at least once, and looking at the Include/Private/Exclude Assets docs numerous times.